kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java

```diff
@@ -0,0 +1,145 @@
+
+    // Create Service Account
+    serviceAccount1 = new ServiceAccountBuilder()
+      .withNewMetadata().withName(SERVICE_ACCOUNT).endMetadata()
+      .build();
+    client.serviceAccounts().inNamespace(currentNamespace).create(serviceAccount1);
+
+    // Bind Impersonator Role to current user
+    impersonatorRoleBinding = new KubernetesClusterRoleBindingBuilder()
+      .withNewMetadata()
+      .withName("impersonate-role")
+      .endMetadata()
+      .addToSubjects(new KubernetesSubjectBuilder()
+        .withApiGroup("rbac.authorization.k8s.io")
+        .withKind("User")
+        .withName(client.currentUser().getMetadata().getName())
+        .withNamespace(currentNamespace)
+        .build()
+      )
+      .withRoleRef(new KubernetesRoleRefBuilder()
+        .withApiGroup("rbac.authorization.k8s.io")
+        .withKind("ClusterRole")
+        .withName("impersonator")
+        .build()
+      )
+      .build();
+
+    client.rbac().kubernetesClusterRoleBindings().inNamespace(currentNamespace).createOrReplace(impersonatorRoleBinding);
+  }
+
+
+  @Test
+  public void testUserImpersonation() {
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(SERVICE_ACCOUNT);
+    requestConfig.setImpersonateGroups("system:authenticated", "system:authenticated:oauth");
+
+    User user = client.currentUser();
+    assertThat(user.getMetadata().getName()).isEqualTo(SERVICE_ACCOUNT);
+  }
+
+
+  @Test
+  public void testCreateProjectImpersonatingServiceAccount() {
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(SERVICE_ACCOUNT);
+    requestConfig.setImpersonateGroups("system:authenticated", "system:authenticated:oauth");
+    // Create a project
+    ProjectRequest projectRequest = client.projectrequests().createNew()
+      .withNewMetadata()
+      .withName(NEW_PROJECT)
+      .endMetadata()
+      .done();
+
+    // Grab the requester annotation
+    String requester = projectRequest.getMetadata().getAnnotations().get("openshift.io/requester");
+    assertThat(requester).isEqualTo(SERVICE_ACCOUNT);
+  }
+
+
+  @After
+  public void cleanup() throws InterruptedException {
+    // Reset original authentication
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(null);
+    requestConfig.setImpersonateGroups(null);
+    client.rbac().kubernetesClusterRoles().inNamespace(currentNamespace).delete(impersonatorRole);
+    client.rbac().kubernetesClusterRoleBindings().inNamespace(currentNamespace).delete(impersonatorRoleBinding);
+    client.projects().delete(new ProjectBuilder().withNewMetadata().withName(NEW_PROJECT).endMetadata().build());
+    client.serviceAccounts().inNamespace(currentNamespace).delete();
+    // Wait for resources to get destroyed
+    Thread.sleep(3000);
```

Comment:
How about using Awaitility https://github.com/fabric8io/kubernetes-client/pull/1121/files#diff-e892e898797ea5223e863136b7348d7cR140 instead?

Reply:
Not sure if that's really needed. I'd rather remove the Thread.sleep now that I see that it doesn't matter when the resources get discarded